trie/transition.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2021 go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package trie
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+)
+
+type TransitionTrie struct {
+	overlay *VerkleTrie
+	base    *SecureTrie
+}
+
+func NewTransitionTree(base *SecureTrie, overlay *VerkleTrie) *TransitionTrie {
+	return &TransitionTrie{
+		overlay: overlay,
+		base:    base,
+	}
+}
+
+func (t *TransitionTrie) Base() *SecureTrie {
+	return t.base
+}
+
+// GetKey returns the sha3 preimage of a hashed key that was previously used
+// to store a value.
+//
+// TODO(fjl): remove this when StateTrie is removed
+func (t *TransitionTrie) GetKey(key []byte) []byte {
+	if t.overlay.GetKey(key) != nil {
+		return t.overlay.GetKey(key)
+	}
+	return t.base.GetKey(key)
+}
```

Comment:
Why we need to do the `t.overlay.GetKey(...)` call if resolving this is only related to a query to the preimages database? Shouldn't we always ask and know we'll get a correct answer from the base tree?

Also, I looked what `GetKey(...)` means for verkle and [I'm not sure that implementation is correct](https://github.com/gballet/go-ethereum/blob/7ec676442f2cc6921bde82f34240bb5e1753eeeb/trie/verkle.go#L57-L59). As in, not really resolving what is expected and actually making the logic here work incorrectly.

Reply:
yup it's all buggy. To be fair, this function isn't used very often, if at all, and there's a proposal by Vitalik to also put the preimage in the C-level polynomial so we never have to worry about converting again. I don't think it makes sense to spend time implementing this function at all, tbh, so I'll replace it with a panic.